source/_posts/2020-07-28-Kubernetes-controllers-A-New-Hope.md

```diff
@@ -0,0 +1,308 @@
+* The proposed ABI mixes medium level and low level. We want to try to implement an high level function in our ABI for watching Kubernetes resources, in order to register a single watch for all the controllers.
+* The actual ABI doesn't have error management at all, but that shouldn't be hard to implement: the serialized ABI data structures might be enums with error as one of the variants, so then the error is propagated back to the user as a `Result`.  
+
+## Wait, where is Golang?
+
+Our _dream_ is to port existing controllers to this approach. Most of Kubernetes world today lives in Golang, but up until this point I didn't talk at all about Golang. Why?
+
+The reason is that we found some critical issues within the Wasm/Golang support, among the others Golang assumes that the Wasm engine runs inside a JS VM using [wasm_exec.js](https://github.com/golang/go/blob/master/misc/wasm/wasm_exec.js). In order to understand the impact of this choice, let me go through the Golang standard library implementation in Wasm.
+
+Today in Golang, when you compile to Wasm, all the standard library I/O operations goes through the [`syscall/js` module](https://golang.org/pkg/syscall/js/), which invokes a predefined set of imports in your module. From the doc: _Package js gives access to the WebAssembly host environment when using the js/wasm architecture. Its API is based on JavaScript semantics._ Going through the documentation should give you the idea of the purpose of this module:
+
+* [`Global()`](https://golang.org/pkg/syscall/js/#Global) allows you to get the `global` object
+* [`Invoke()`](https://golang.org/pkg/syscall/js/#Value.Invoke) allows you to use a `js.Value` as a function and invoke it
+
+Now let's analyze how the `net/http` client implementation on Wasm: When the user invokes `RoundTrip`, the request [is transformed in a Javascript object](https://github.com/golang/go/blob/b56791cdea5caa87ffcd585d29c294bd3d08a06a/src/net/http/roundtrip_js.go#L60), then using the syscall `Global` and `Call` the javascript global function [`fetch` is invoked](https://github.com/golang/go/blob/b56791cdea5caa87ffcd585d29c294bd3d08a06a/src/net/http/roundtrip_js.go#L106) and finally, on the returned promise, the [two callbacks are set](https://github.com/golang/go/blob/b56791cdea5caa87ffcd585d29c294bd3d08a06a/src/net/http/roundtrip_js.go#L167) to handle errors and responses.
+
+The problem with this approach, although it works well in the browsers, is that the **engine needs to implement all these JS functions used by Golang standard library**. Somehow, the engine should fake itself as a Javascript engine implementing the Javascript standard library. Not only, it requires to **secure access** to the _global_ object fields, in order to retain the isolation properties. Making `syscall/js` optional could make far more doable implementing controllers in Golang using Wasm.
```

Comment:
```suggestion
The problem with this approach, although it works well in the browsers, is that the **engine needs to implement all these JS functions used by Golang standard library**. If we're not using a Javascript based runtime, we need to reimplement this. Further, it requires us to **secure access** to the _global_ object fields, in order to retain the isolation properties.
```

Reply:
Why removing the last sentence? because it's a repetion of something i underline later?